nntrainer/layers/gather_layer.cpp

```diff
@@ -0,0 +1,109 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2025 SeungBaek Hong <sb92.hong@samsung.com>
+ *
+ * @file   gather_layer.cpp
+ * @date   02 April 2025
+ * @see    https://github.com/nnstreamer/nntrainer
+ * @author SeungBaek Hong <sb92.hong@samsung.com>
+ * @bug    No known bugs except for NYI items
+ * @brief  This is gather layer class (operation layer)
+ */
+
+#include "common_properties.h"
+#include <gather_layer.h>
+#include <nntrainer_error.h>
+#include <nntrainer_log.h>
+#include <node_exporter.h>
+#include <stdexcept>
+#include <util_func.h>
+
+#include <layer_context.h>
+
+namespace nntrainer {
+
+void GatherLayer::finalize(InitLayerContext &context) {
+  axis = std::get<props::Axis>(gather_props).get();
+  TensorDim inputDim = context.getInputDimensions()[0];
+  TensorDim indexDim = context.getInputDimensions()[1];
+
+  if (axis < 1 || axis > 3) {
+    throw std::invalid_argument(
+      "The axis property of GatherLayer should be between 1 and 3.");
+  }
+
+  if (inputDim[0] != indexDim[0]) {
+    throw std::invalid_argument(
+      "The batch size of the input and index should be same.");
+  }
+
+  context.setOutputDimensions({indexDim});
+}
+
+void GatherLayer::forwarding_operation(const Tensor &input, const Tensor &index,
+                                       Tensor &output) {
+  for (unsigned int b = 0; b < index.getDim().batch(); ++b) {
+    for (unsigned int i = 0; i < index.getDim().channel(); ++i) {
+      for (unsigned int j = 0; j < index.getDim().height(); ++j) {
+        for (unsigned int k = 0; k < index.getDim().width(); ++k) {
+          auto selected = (size_t)index.getValue(b, i, j, k);
+          if (selected >= input.getDim()[axis]) {
```

Comment:
I think it will take so many time. We do need optimize this later. please leave the note. And I wonder what is the difference with concat layer?

Reply:
@jijoongmoon there is no any common ground between concat and gather...

For a 3-D tensor the output is specified by:
out[i][j][k] = input[index[i][j][k]][j][k]  # if dim == 0
out[i][j][k] = input[i][index[i][j][k]][k]  # if dim == 1
out[i][j][k] = input[i][j][index[i][j][k]]  # if dim == 2

gather example 1:
```
>>> t = torch.tensor([[1, 2], [3, 4]])
>>> indices = torch.tensor([[0, 0], [1, 0]])
>>> torch.gather(t, 1, indices)
tensor([[ 1,  1],
        [ 4,  3]])
```

gather example 2:
```
>>> t = torch.tensor([[1, 2, 3], [4, 5, 6], [7, 8, 9]])  
>>> indices = torch.tensor([[0, 2], [0, 1], [0, 2]])
>>> torch.gather(t, 0, indices)
tensor([[1, 8],
        [1, 5],
        [1, 8]])
```
ref: https://pytorch.org/docs/stable/generated/torch.gather.html